### ISSUE TITLE
Use underscores everywhere on internals

### ISSUE THREAD

**Issue by rafaqz** [JuliaObjects/ConstructionBase.jl#94]
Even with the new public keywords we need to make internals look scary, the consequences of changing them is too large now.

See #93

**Comment by jw3126** [comment]
I don't think it is our responsibility to prefix every private function. If we really want to make private stuff private, we could use the following pattern:
```julia
module ConstructionBase

function setproperties end
function getproperties end
...

module _Private
import ..ConstructionBase: setproperties, getproperties, ...

include("some.jl")
include("files.jl")
...
end
```

**Comment by rafaqz** [comment]
I'm not sure I see the reasoning, underscores are  less fuss and everyone knows what they mean.

And most of my packages use underscores and a lot of Base Julia does, it's not an unusual thing to do.

**Comment by jw3126** [comment]
Sure, it is about the "everywhere" part. Typically some internal functions are prefixed with `_` but it requires a lot of discipline to do it with every private function. Will I have often seen some internals beeing prefixed, I have never seen it being done fully consistently.

**Comment by rafaqz** [comment]
I'm just thinking about the relative scale of that discipline vs the massive ecosystem outages confusion about our methods can cause.

**Comment by jw3126** [comment]
> I'm just thinking about the relative scale of that discipline vs the massive ecosystem outages confusion about our methods can cause.

Prefixing every function with `_` vs putting them into a module both cause the same outage. For backwards compat you can do:
```julia
module ConstructionBase
function getproperties end
function setproperties end
...
# Internal functions that are de facto used by other packages however
# we keep them here to not break these packages
function setproperties_object end
...

```

**Comment by rafaqz** [comment]
Yes you can do the same thing either way. I'm just not sure why we would do something different to what Base julia and most people do, which is use underscores. But if you really want separate modules sure.

The idea of this post was to respond constructively to breaking half of the SciML ecosystem.

**Comment by jw3126** [comment]
> Yes you can do the same thing either way. I'm just not sure why we would do something different to what Base julia and most people do, which is use underscores. But if you really want separate modules sure.
> 
> The idea of this post was to respond constructively to breaking half of the SciML ecosystem.

I read your opening post as a higher bar than what base julia or other packages do. Not all internals of Base are prefixed, internals of Base get used all the time and nightly breaks some ecosystem every day or so.

My understanding was you want to be fully rigorous about what is private. In this scenario my suggestion was to 
enforce it through code rather than rely on discipline.

**Comment by jw3126** [comment]
Anyway if you make a PR that adds underscores everywhere I am not going to object.

**Comment by rafaqz** [comment]
This package is very small, so the ratio of dependencies to lines of code is actually *much* higher than Base Julia! So I would argue effort/reward for things like this is higher here.

See when I have time for the PR, but I think we should also deprecate the non-underscore functions so people get a warning to stop using them.
